poly/schnorr.go

```diff
@@ -0,0 +1,270 @@
+// You call this function when you want a new signature to be issued on a specific message
+// The security of the d. schnorr signature protocol is the same as the regular :
+// The random secret MUST BE FRESH for EACH signature / signed message (hence the 'NewRound')
+func (s *Schnorr) NewRound(random *SharedSecret, msg []byte) error {
+	s.random = random
+	s.hash = nil
+	s.partials = nil // erase the previous partil signature from previous round
+	s.partials = make([]*PartialSchnorrSig, s.info.N)
+	hash, err := s.hashMessage(msg, s.random.Pub.SecretCommit())
+	if err != nil {
+		return errors.New(fmt.Sprintf("Unable to hash the message with the given shared secret : %v", err))
+	}
+	s.hash = &hash
+	return nil
+}
+
+// hash returns a hash of the message and the random secret
+// H( m || V )
+// Returns an error if something went wrong with the marshalling
+func (s *Schnorr) hashMessage(msg []byte, v abstract.Point) (abstract.Secret, error) {
+	vb, err := v.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	c := SUITE.Cipher(vb)
+	c.Message(nil, nil, msg)
+	return SUITE.Secret().Pick(c), nil
+}
+
+// Verify if the received structures are good
+// tests the partials shares if there is some
+func (s *Schnorr) verify() error {
+	if s.Longterm.Index != s.random.Index {
+		return errors.New("The index for the longterm shared secret and the random secret differs for this peer.")
+	}
+	nsig := 0
+	for i, _ := range s.partials {
+		if s.partials[i] != nil {
+			nsig += 1
+		}
+	}
+	if nsig < s.info.T {
+		return errors.New(fmt.Sprintf("Received to few Partial Signatures (%d vs %d) to complete a global schnorr signature", len(s.partials), s.info.T))
+	}
+	return nil
+}
+
+// verifyPartialSig will verify if a given partial signature can be checked against the longterm and random secrets
+// of the schnorr structures.
+func (s *Schnorr) verifyPartialSig(ps *PartialSchnorrSig) error {
+	// compute left part of the equation
+	left := SUITE.Point().Mul(SUITE.Point().Base(), *ps.Part)
+	// compute right part of the equation
+	right := SUITE.Point().Add(s.random.Pub.Eval(ps.Index), SUITE.Point().Mul(s.Longterm.Pub.Eval(ps.Index), *s.hash))
+	if !left.Equal(right) {
+		return errors.New(fmt.Sprintf("Partial Signature of peer %d could not be validated.", ps.Index))
+	}
+	return nil
+}
+
+// index returns the index of the peer holding this schnorr struct
+// the index of its share in the polynomials used
+func (s *Schnorr) index() int {
+	return s.Longterm.Index
+}
+
+// RevealPartialSig reveals the partial signature for this peer
+// Si = Ri + H(m || V) * Pi
+// with :
+// 	- Ri = share of the random secret for peer i
+//  - V  = public commitment of the random secret (i.e. Public random poly evaluated at point 0 )
+//  - Pi = share of the longterm secret for peer i
+// This signature is to be sent to each others peers
+func (s *Schnorr) RevealPartialSig() *PartialSchnorrSig {
```

Comment:
For this one, since it is a **public** function I prefer the more explicit terms instead of the shorter one...